### ISSUE TITLE
Actioncable deadlock resuming a connection with multiple channels

### ISSUE THREAD

**Issue by eacaps** [rails/rails#24094]
I ran into an issue with my ActionCable enabled app where the server was locking up constantly using the current master branch from the rails repository. I was able to recreate the issue in my own fork of the ActionCable examples repo.

I modified the ActionCable example to mount ActionCable at the '/ws' endpoint in the routes.rb file and I also created a second channel which the client subscribes to at the same time as the first: https://github.com/rails/actioncable-examples/compare/master...eacaps:2channel-rails5-master. 

### Steps to reproduce

```
Open a terminal
$ git clone https://github.com/eacaps/actioncable-examples.git
$ cd actioncable-examples
$ git checkout 2channel-rails5-master
$ ./bin/setup
$ ./bin/rails server
Open http://localhost:3000/ in a browser
Click on a name to log in
```

After you click on a name to log in the server seems to lock up. Attempting to follow the navigation will result in hung requests. 

### Expected behavior
I created a separate branch that uses `5.0.0beta3` called `2channel-beta3`. If you checkout that code and run the instructions above the server responds fine and your terminal output will look more like this:

```
Started GET "/ws" for 127.0.0.1 at 2016-03-07 10:46:14 -0500
Started GET "/ws/" [WebSocket] for 127.0.0.1 at 2016-03-07 10:46:14 -0500
Successfully upgraded to WebSocket (REQUEST_METHOD: GET, HTTP_CONNECTION: Upgrade, HTTP_UPGRADE: websocket)
  User Load (0.1ms)  SELECT  "users".* FROM "users" WHERE "users"."id" = ? LIMIT ?  [["id", 3], ["LIMIT", 1]]
[ActionCable] [Funkmaster Flex] Registered connection (Z2lkOi8vYWN0aW9uY2FibGUtZXhhbXBsZXMvVXNlci8z)
i subscribed[ActionCable] [Funkmaster Flex] CommentsChannel is transmitting the subscription confirmation

[ActionCable] [Funkmaster Flex] OtherChannel transmitting {:data=>"some data"}
[ActionCable] [Funkmaster Flex] OtherChannel is transmitting the subscription confirmation
[ActionCable] [Funkmaster Flex] CommentsChannel#unfollow
```

Everything will work just fine from here.

### Actual behavior
When using the `2channel-rails5-master` branch your terminal will probably look something like this:

```
Started GET "/ws" for 127.0.0.1 at 2016-03-07 10:36:09 -0500
Started GET "/ws/" [WebSocket] for 127.0.0.1 at 2016-03-07 10:36:09 -0500
Successfully upgraded to WebSocket (REQUEST_METHOD: GET, HTTP_CONNECTION: Upgrade, HTTP_UPGRADE: websocket)
  User Load (0.4ms)  SELECT  "users".* FROM "users" WHERE "users"."id" = ? LIMIT ?  [["id", 1], ["LIMIT", 1]]
[ActionCable] [The Notorious B.I.G.] Registered connection (Z2lkOi8vYWN0aW9uY2FibGUtZXhhbXBsZXMvVXNlci8x)
```

The server is hung and there doesn't seem to be a way to recover.

### System configuration
Using latest rails master branch from:
`gem 'rails', git: 'https://github.com/rails/rails.git', branch: 'master'`

Ruby version info:
```
ruby -v
ruby 2.2.2p95 (2015-04-13 revision 50295) [x86_64-darwin15]
```

### Additional info
I also created `1channel-beta3` and `1channel-rails5-master` branches and they seem unaffected by this issue so in my limited research it seems that it has something to do with subscribing to multiple channels simultaneously. Like I show above, it also seems to work with rails `5.0.0beta3` unfortunately that release has an issue that prevents deployment on Heroku. That issue is fixed in the master branch unfortunately I ran into this separate issue. It can be temporarily remedied by creating a master channel that will subscribe to multiple separate channels but I don't think that is ideal going forward

**Comment by AndreasWurm** [comment]
encountered the same issue, nice work making a reproducible setup

**Comment by maclover7** [comment]
cc @matthewd, this might have something to do with Active Record connection management...?

**Comment by eacaps** [comment]
I finally tracked this one down, when multiple channel subscriptions come in at the same time the **add** method in action_cable/connection/subscriptions.rb attempts to load the channel classes on line 29: 

`subscription_klass = connection.server.channel_classes[id_options[:channel]]`

that runs the following code in action_cable/server/base.rb 

```
      def channel_classes
        @channel_classes || @mutex.synchronize do
          @channel_classes ||= begin
            config.channel_paths.each { |channel_path| require channel_path }
            config.channel_class_names.each_with_object({}) { |name, hash| hash[name] = name.constantize }
          end
        end
      end
```

The first async thread in grabs the server @mutex and begins loading the channels. When the second channel subscription begins processing, it runs the same code but is blocked because it cannot get the @mutex, thus deadlocking the actioncable server.

The simplest fix I came up with was just loading the channel_classes in the **initialize** method of action_cable/server/base.rb. Anyone have a better idea?
